bbp/src/gp/getpar/src/getarg.3
.\" @(#)getarg.3	56.1 10/25/93
.nr Np 0 1
.TH GETARG 3 "May 1991"
.SH NAME
.B getarg 
routines -\ get arguments from a string
.br
.SH SYNOPSIS
#include "libget.h"
.PP
.B void setarg(char *argstring, char *subname)
.PP
.B void setarg_(char *argstring, char *subname,
.B int dum1, int dum2)
.PP
.B int countarg(char *name, char *type)
.PP
.B int cntarg_(char *name, char *type,
.B int dum1, int dum2)
.PP
.B int lenarg(char *name)
.PP
.B int getarg(char *name, char *type,
.B void *pointer)
.PP
.B fgtarg_(char *name, char *type,
.B void *pointer, int dum1, int dum2, int lens)
.PP
.B void endarg(void)
.PP
.B void endarg_(void)
.PP
In the subroutines 
.B getarg()
and
.B fgtarg_()
the declaration for the argument
.B pointer
depends on the argument type
.B type.
.br
.SH DESCRIPTION
.br
The
.B getarg
routines provide a simple procedure for passing optional arguments
to subroutines and for parsing strings into data values.  The syntax
of the functions and the format of the parameter specification
is similar to
.B getpar(3).
.PP
The subroutine
.I setarg()
initializes the
.B getarg
package, and it must be called before any other references to
.B getarg.
The two arguments to
.I setarg() 
are a character string containing optional arguments
and a subroutine name (for error reporting).
The pointer to either the character string or the subroutine name may be
.I NULL
but they cannot both be
.I NULL.
.PP
The subroutine
.I countarg()
takes two arguments:
.I name
and
.I type.
.I Name
is the external name of the variable initialized in
.I setarg().
.I Type
specifies the variable type, one of 
.I d, f, F, 
or 
.I s.
.I Countarg
counts the number of elements for the vector
.I name.
It is useful for ascertaining the vector limit
for data retrieval via
.I getarg().
Upon successful completion,
.I countarg()
returns the number of elements for the parameter, or zero if there are no
elements.
.PP
The subroutine
.I lenarg()
takes one argument:  
.I name.
.I Name
is the external name of the variable initialized in
.I setarg().
.I Lenarg()
determines the length of the longest element in the
.I name
vector.
It is useful for ascertaining the memory space needed to
accommodate the largest string in a character string vector.
Upon successful completion,
.I lenarg()
returns a value guaranteed to be at least as large as the 
length of the longest string argument.  It returns zero
if there are no arguments.
.PP
The subroutine
.I getarg()
takes three arguments:
.I name, type,
and
.I pointer.
.I Name
is the external name of the variable initialized in
.I setarg().
.I Type
specifies the variable type.
The currently valid types are:
.IP "" 5
"d"	integer
.br
"f"	float
.br
"F"	double
.br
"s"	character string
.br
"b"	boolean (integer)
.br
"vd"	integer vector
.br
"vf"	float vector
.br
"vF"	double vector
.br
"vs"	character string vector
.PP
The 
.I type
parameter can be expanded to indicate the maximum number of elements
allowable in vectors.  For example type="vf[4]" or type="vf(4)"
would cause
.I getarg()
to modify no more than 4 elements of the vector, regardless of how
many elements are in the argument list
initialized in
.I setarg().
If no limit is specified, a limit of 10 is quietly enforced.  If there 
are fewer elements than the limit the vector will only be modified up
to the number of elements.
.PP
.I Pointer
is a pointer to the type of variable indicated by
.I type.
If no occurrences of the parameter
.I name
is found, the contents of
.I pointer
are not modified.
.PP`
.I Getarg()
returns 1 for a successful conversion, and 0 if unsuccessful.  In the
case of vectors
.I getarg()
returns the number of elements found, or 0 if an unsuccessful conversion
occurs at any index in the vector.
.PP
The subroutine
.I endarg()
gracefully terminates the package.
.PP
The Fortran interface subroutines
.I setarg_()
,
.I cntarg_()
,
.I fgtarg_()
and
.I endarg_()
are equivalent with the exception that 
.I fgtarg_() 
does not process vectors of character strings.  Also, there is no
Fortran equivalent for
.I lenarg().
.PP
The following program example illustrates the use of 
.B getarg
for the purpose of parsing optional subroutine arguments.
It is intended for use with subroutines that have a few essential
arguments and many optional ones.
.IP "" 5
sub(x,lx,list)
.br
float *x;
.br
int lx; 
.br
char *list;
.br
   {
.br
	int opt1, boo;
.br
	float fopt2;
.br
	char title[40];
.br

.br
	/* set defaults */
.br
	opt1= 10;
.br
	fopt2= 5.0;
.br
	title[0]= '\0';
.br
	boo= 1;	/* true */
.br

.br
	setarg(list,"sub");	/* initialize getarg package */
.br
	getarg("opt1","d",&opt1);
.br
	getarg("fopt2","f",&fopt2);
.br
	getarg("title","s",title);
.br
	getarg("boo","b",&boo);
.br
	endarg();		/* deactivate getarg package */
.br

.br
	/* rest of subroutine */
.br
   }
.PP
The minimal call to this subroutine would be
.IP "" 5
sub(x,lx,0);
.br
.ti 0
or
.br
sub(x,lx,"");
.PP
in which case the default values of all the options are used.
To change some of the options, the call would be
.IP "" 5
sub(x,lx,"opt1=100 fopt2=2.0 title='new title' noboo");
.br
.PP
.B NOTE:
If the variables are defaulted by giving them initializations in
static declarations,
then subsequent calls to subroutines that use
.B getarg
may have their default values modified by previous calls.
.PP
The next program example illustrates the use of
.B getarg 
for the purpose of parsing a string argument into data values.
.IP "" 5
get_msg(list)
.br
char *list;
.br
   {
.br
	double *time;
.br
	char **sta_chans, dbtemp[BUFSIZ], format[BUFSIZ];
.br
	int orid, count_times, count_stas;
.br
	int get_times, get_stas, len_stas;
.br

.br
	setarg(list, NULL); /* initialize getarg package */
.br

.br
	count_times = countarg("channel-time", "F");  /* determine size of double vector */
.br
	time = (double *) malloc(count_times * sizeof(double)); /* create space */
.br
	sprintf(format, "vF[%d]", count_times); /* format the type parameter */
.br
	get_times = getarg("channel-time", format, time);  /* retrieve the data */
.br
	if (count_times != get_times)
.br
	  	/* error */
.br

.br
	/*
.br
	 * For the string vector ascertain the number of elements and
.br
	 * the length of the longest element in the vector channel.
.br
	 * Malloc the appropriate amount of memory space before calling
.br
	 * getarg().
.br
	 */
.br
	count_stas = countarg("channel", "s");
.br
	len_stas = lenarg("channel");
.br
	sta_chans = (char **) malloc(count_stas * sizeof(char*));
.br
	for (i = 0; i < count_stas; i++)
.br
		sta_chans[i] = (char *) malloc((len_stas + 1) * sizeof(char));
.br
	sprintf(format, "vs[%d]", count_stas);
.br
	get_stas = getarg("channel", format, sta_chans);
.br
	if (count_stas != get_stas)
.br
		/* error */
.br

.br
	/* get the remaining arguments */
.br
	getarg("orid", "d", &orid);
.br
	getarg("dbtemp", "s", dbtemp);
.br

.br
	endarg(); /* deactivate getarg package */
.br

.br
	/* rest of subroutine */
.br
   }
.PP
An example call to this subroutine would be
.IP "" 5
get_msg("dbtemp='demo/demo' orid=721 channel-time=240.0,480.0 channel='ARA/sz','ARA/bz')
.SH PARAMETER FORMAT
The parameters in the strings can occur in any order, and any number
of times.  In the case of multiple specifications, the last one is
used.  Any parameters that are not requested by
.I getarg()
are ignored.
.PP
Each specification is of the form 
.I name=value.
No embedded blanks are allowed on either side of the equals ("=")
sign.  Character strings with white space are delimited with single
(') or double (") quotes which are removed by
.I getarg().
To get a single or double quote in a string, precede it with a back-slash (\\).
Vector arguments are separated by commas.  Vectors of strings are
enclosed in single quotes separated by commas.  To get a single 
quote, comma, or back-slash in a string argument, precede each with a 
back-slash.  For example:
.IP "" 5
get_msg("vs_arg='The Dommermuth\\'s','619\\\\234-3153','San Diego\\, CA'");
.PP
The only exception to the
.I name=value
syntax is boolean variables which can simply be specified by
.I name
or
.I noname
to mean true or false respectively.
.PP
Repetition factors can be used to specify repeated values in vectors
using the symbols
.I x, X, and *
interchangeably.  For example
.I get_msg("ints=1,2x2,3,3*4,2X5")
would yield a vector with nine elements:  1, 2, 2, 3, 4, 4, 4, 5, 5.
Repetition factors cannot be used with vectors of strings.  
.SH AVAILABILITY
.B getarg
routines are in the library
.I /opt/util/lib/libget.a
which can be loaded with
.I -lget
provided the
.I -L/opt/util/lib
path is specified.  The include file libget.h can be found in
/opt/util/include.
.SH "SEE ALSO"
getpar(3)
.SH DIAGNOSTICS
The routines produce error messages on
.I stderr
and return -1 for the following trapped errors.
.PP
1.  
.I  setarg()
is not (properly) called.
.br
2.  The arglist to
.I setarg()
appears to be garbage.
.br
3.  
.I  countarg(),
.I  lenarg(),
or
.I getarg()
are called before
.I setarg().
.br
4.  
.I  countarg(),
.I lenarg(),
or
.I getarg()
are called after
.I endarg().
.br
5.  There is an unknown conversion type in
.I countarg()
or
.I getarg().
.br
6.  Attempt to process a vector of strings in
.I fgtarg_().
.br
7.  Invalid vector limit is specified in
.I getarg().
.br
8.  Invalid vector repetition factor is specified in
.I getarg().
.br
9.  A NULL
.I pointer
parameter is passed to
.I getarg().
.PP
An attempt is made to report which subroutine was active when the error occurred.
.SH BUGS
Probably.
.SH AUTHOR
Robert W. Clayton, Seismological Laboratory, Caltech, Pasadena, CA
91125
.br
Cynde K. Smith, Science Applications International Corp., San Diego,
CA  92121


doc/RELEASE_NOTES
List of TRAC features/issues implemented
========================================

BBP - 16.5.0

Trac #298 - Define evaluation criteria for comparing BBP Releases
Trac #302 - Fix matplotlib deprecation warnings for versions 1.3 and greater
Trac #303 - Integrate RMG rupture generator module into BBP Platform
Trac #304 - Update GP rupture generator code to version 5.0
Trac #305 - Update BBToolbox to version 1.6
Trac #309 - Update GP Rupture Generator to version 5.0.1
Trac #310 - Update GP site response module to version 2014
Trac #312 - Fix compilation warnings for various codes in the GP method
Trac #313 - Modify cluster scripts to use scec-00 filesystem
Trac #314 - Include parameters from RZZ2015 into BBP workflows
Trac #315 - Include parameters from AS2016 in the BBP Workflows
Trac #318 - Add PyProj version to software_status file
Trac #325 - Respect module fails to run for observed seismograms
Trac #326 - Add Anderson GOF to Broadband Platform
Trac #328 - Update LABasin, Mojave and NoCal GFs
Trac #329 - Add RotD100/RotD50 ratio metric to BBP

BBP - 15.3.0

Trac #158 - Platform should write simulation results on BBP_DATA_DIR
Trac #200 - Add version number to BBP XML workflow description file
Trac #207 - Use real station distances in platform
Trac #227 - Convert UCSB method to use GNU compilers
Trac #228 - Convert SDSU method to use GNU compilers
Trac #239 - Add the NGA-West2 GMPE models to the Broadband Platform
Trac #261 - Make simulation inputs available in data output directory
Trac #262 - Add configurable line width option for BBP seismogram plots
Trac #263 - Include 3 GMPE models in the Broadband Platform
            for use in the East
Trac #264 - Enable LF seismograms module to accept input directory
Trac #265 - Update Match module to work with user-provided seismograms
Trac #266 - Update station list generator from PyNGA to allow for
            footwall and hanging wall stations
Trac #267 - Remove display requirement from station generation script
Trac #270 - Fix header in rd50 files in outdata/obs_seis
            directory in GMPE runs
Trac #271 - Integrate the UCSB 14.8 code into BBP
Trac #274 - Add option to validation cluster script to run only
            rupture generator
Trac #275 - Allow users to see missing velocity models when
            selecting a validation event
Trac #276 - Add random SEED used in simulations to HTML summary in outdata
Trac #277 - Copy SRC and method-specific parameter files to outdata
Trac #278 - Integrate SDSU BBToolbox v1.5.5.1 into Broadband Platform
Trac #279 - Run BBP on cluster using the same command on all nodes
Trac #280 - Add rule to GP method to calculate DX/DY automatically
Trac #281 - Label BBP methods that are not yet validated
Trac #282 - Update SDSU BBtoolbox to fix bug when calculating
            seismograms at >70km
Trac #283 - Modify directory structure in validation packages to add
            common input files directory
Trac #285 - Error parsing XML file created by BBP
Trac #286 - ExSim results incorrect for some close by stations
Trac #287 - Removed unused or obsolete codes and data files from Platform
Trac #288 - Increase array size in RotD50 code to avoid truncating timeseries
Trac #289 - Integrate Jan 2015 version of the UCSB source and wave
            propagation code
Trac #290 - Update BBP distribution directory structure
Trac #291 - Improve BBP command-line interface with more intuitive questions
Trac #292 - Rename LOMAP velocity model to NOCAL
Trac #293 - Include method name on html report in the outdata directory
Trac #294 - Update SDSU codebase with latest February 2015 changes
Trac #295 - Add support for varying rupture velocity in GP method
Trac #299 - README references icc compilers
Trac #300 - Add expert mode to run_bbp.py
Trac #301 - Fix units tests test_vm2vm and test_s2v

BBP 14.3.0

Trac # 114 - Convert from f77 to gfortran
Trac # 146 - Move DT out of src file
Trac # 195 - Produce GMPE plot for validation simulations
Trac # 212 - wrong unit on respect plot
Trac # 213 - Remove unneeded GP_GOF parameters
Trac # 214 - Reduce volume of logged data
Trac # 215 - Modify SRF plot so that it uses same ALONG_STRIKE/DOWN_DIP
             reference as SRC file
Trac # 219 - SRC file should specify Mw, platform then calculates
             the seismic moment
Trac # 222 - Integrate new GP Rupture Generator v3.2.1
Trac # 224 - Add randomization in BBToolbox using the iseed parameter
Trac # 230 - Modify station list parser to accept float vs30 values
Trac # 232 - Add the decimation factor patch from SDSU to BBToolbox
Trac # 233 - Show epicenter location on all map plots
Trac # 234 - Add '08 suffix to current GMPE plot labels
Trac # 235 - Modify GoF plot labels
Trac # 237 - Rotate all multi-figure plots so they are viewed in landscape
Trac # 240 - Run LF and HF components in the UCSB method using separate GFs
Trac # 241 - Modify SDSU BBtoolbox to handle smaller-magnitude events
Trac # 242 - Integrate GP Rupture generator v3.3 into the Broadband Platform
Trac # 243 - Update GP jbsim LF program
Trac # 244 - Arias duration module fails when processing zeroed
             observation files
Trac # 245 - Make str_fac a region-specific parameter in BBToolbox
Trac # 246 - Capture and store resource information from shell
Trac # 247 - Allow user to specify hypocenter randomization area
             in cluster script
Trac # 248 - Add option to cluster script for user to run only the
             rupture generator
Trac # 249 - Add option to cluster script for users to specify SRF files to use
Trac # 250 - Correct NGA model labels on GMPE box plots
Trac # 251 - Integrate ExSIM14 into Broadband Platform
Trac # 252 - Integrate new version of GP high frequency code
Trac # 253 - Simplify makefiles and remove support for user-specified compilers
Trac # 254 - Integrate Feb 2014 version of UCSB code into Broadband Platform
Trac # 256 - Integrate SDSU model version 1.5.4.1 into the Broadband Platform

bbp/src/gp/getpar/src/getarg.c
/* 
 * NAME
 *	getarg
 *
 * DESCRIPTION
 *	get subroutine arguments from a string.  Acquired from Caltech.
 *
 *	See getarg.3 for details.
 *
 * AUTHOR
 * copyright (c) Robert W. Clayton
 *		 Seismological Laboratory
 *		 Caltech
 *		 Pasadena, CA 91125
 *
 * 24 Jul 1991  J Given         Added function getsarg() for dynamic
 *                              string allocation
 *
 * 25 Apr 1991	Cynde K. Smith	Added vector of string capability for C
 *			       	version only.
 * 26 Apr 1991	Cynde K. Smith  Replaced calls to atof and atoi with
 *			       	strtod and strtol to allow error checking.
 * 09 May 1991  Cynde K. Smith  Added countarg() and cntarg_() subroutines.
 * 14 May 1991  Cynde K. Smith  Added lenarg() subroutine.
 * 
 * Getarg routines:
 *
 * Externally visable routines:
 *
 *		setarg(argc,argv)
 *		countarg(name,type)
 *		lenarg(name)
 *		getarg(name,type,valptr)
 *		endarg()
 *
 * To get C-version:
 *		cc -c getarg.c
 *
 * To get F77-version:
 *		cp getarg.c fgetarg.c
 *		cc -c -DFORTRAN fgetarg.c
 *		rm fgetarg.c
 *
 *SccsId: @(#)getarg.c	56.1 10/25/93
 */
#include	<stdio.h>
#include	<stdlib.h>
#include	<stdarg.h>
/*
#ifndef hpux 
#include	<floatingpoint.h>
#endif
*/
#include	<string.h>
#include	"libget.h"

#define MAXNAME		256	/* max length of name */
#define MAXVALUE	5120	/* max length of value */
#define MAXVECTOR	20	/* max # of elements for unspecified vectors */
#define GETARG_ERROR	-1	/* error status for getarg error */

#define INIT	 1	/* bits for FLAGS (ext_arg.argflags) */
#define END_PAR	 2

#define LISTINC		32	/* increment size for arglist */
#define BUFINC		1024	/* increment size for argbuf */

struct arglist		/* structure of list set up by setarg */
   {
	char *argname;
	char *argval;
	int hash;
   };
struct ext_arg		/* global variables for getarg */
   {
	char *progname;
	int argflags;
	struct arglist *arglist;
	struct arglist *arghead;
	char *argbuf;
	int nlist;
	int nbuf;
	int listmax;
	int bufmax;
   }	ext_arg;

/* abbreviations: */
#define AL 		struct arglist
#define PROGNAME	ext_arg.progname
#define FLAGS		ext_arg.argflags
#define ARGLIST		ext_arg.arglist
#define ARGHEAD		ext_arg.arghead
#define ARGBUF		ext_arg.argbuf
#define NLIST		ext_arg.nlist
#define NBUF		ext_arg.nbuf
#define LISTMAX		ext_arg.listmax
#define BUFMAX		ext_arg.bufmax

static int ga_add_entry(char *name, char *value);
static int ga_getarg_err(char *subname, char *format, ...);
static int ga_compute_hash(char *s);
static int ga_getvector(char *list, char *type, void *val);

void
#ifdef FORTRAN
setarg_(char *list, char *subname, int dum1, int dum2)
#else
setarg(char *list, char *subname)
#endif
   {
	register char *pl, *pn, *pv;
	register int i;
	int      lenlist;
	char t, name[MAXNAME], value[MAXVALUE];
	

	PROGNAME= subname;
	FLAGS= INIT;

	ARGLIST = NULL;
	ARGBUF = NULL;
	NLIST= NBUF= LISTMAX= BUFMAX= 0;

#ifdef 	FORTRAN
	lenlist = dum1;
#else
	lenlist = strlen(list);
#endif	
	if(lenlist == 0) return;
	if(list == NULL) return;

	pl= list;
	/* loop over entries on each line */

	for(i=0; i<lenlist && *pl != '\0'; i++)
	{
		while(*pl==' ' || *pl=='\t') pl++;
		if(*pl=='\0'|| *pl=='\n') continue;
		
		/* get name */
		pn= name;
		while(*pl != '=' && *pl != '\0' && *pl != ' '
		      && *pl != '\t') *pn++ = *pl++;
		*pn = '\0';
		if(*pl == '=') pl++;


		/* get value */

		*value= '\0';
		pv= value;
		if(*pl == '\"' || *pl == '\'')
		{
			t = *pl++;
			while(*pl != '\0')
			{
				if(*pl == t)
				{
					   if (pl[-1] != '\\' &&
					       (pl[1] == ' ' || pl[1] == '\0'))
					     {
					       pl++;
					       break;
					     }
				}
				*pv++ = *pl++;
			}
		}
		else	
		{
			while(*pl && *pl != ' ' && *pl != '\t') 
				*pv++ = *pl++;
		}
		
		*pv= '\0';
		if (ga_add_entry(name,value) == GETARG_ERROR)
		    return;
	}
}

/* add an entry to arglist, expanding memory if necessary */
int ga_add_entry(char *name, char *value)
   {
	struct arglist *alptr;
	int len;
	register char *ptr;

	/*fprintf(stderr,"getarg: adding %s (%s)\n",name,value);*/
	/* check arglist memory */
	if(NLIST >= LISTMAX)
	   {
		LISTMAX += LISTINC;
		if(ARGLIST == NULL)
			ARGLIST= (AL *)malloc(LISTMAX * sizeof(AL));
		 else	ARGLIST= (AL *)realloc(ARGLIST,LISTMAX * sizeof(AL));
	   }
	/* check argbuf memory */
	len= strlen(name) + strlen(value) + 2; /* +2 for terminating nulls */
	if(NBUF+len >= BUFMAX)
	   {
		BUFMAX += BUFINC;
		if(ARGBUF == NULL)
			ARGBUF= (char *)malloc(BUFMAX);
		 else	ARGBUF= (char *)realloc(ARGBUF,BUFMAX);
	   }
	if(ARGBUF == NULL || ARGLIST == NULL)
		return ga_getarg_err("setarg","cannot allocate memory");

	/* add name */
	alptr= ARGLIST + NLIST;
	alptr->hash= ga_compute_hash(name);
	ptr= alptr->argname= ARGBUF + NBUF;
	do *ptr++ = *name; while(*name++);

	/* add value */
	NBUF += len;
	alptr->argval= ptr;
	do *ptr++ = *value; while(*value++);
	NLIST++;
	return 0;
   }

void 
#ifdef FORTRAN
endarg_()
#else
endarg(void) /* free arglist & argbuf memory, & process STOP command */
#endif
   {
	if(ARGLIST != NULL) free(ARGLIST);
	if(ARGBUF  != NULL) free(ARGBUF);
	ARGBUF=  NULL;
	ARGLIST= NULL;
	FLAGS= END_PAR;	/* this stops further getarg calls */
   }


/* count the number of arguments for a particular parameter name */

int
#ifdef FORTRAN
cntarg_(char *name, char *type, int dum1, int dum2)
#else
countarg(char *name, char *type)
#endif
{
	int 	found, h;
	char 	*str, *ptr;
	register struct arglist *alptr;
	
	if (FLAGS & END_PAR)
	{
		return ga_getarg_err("countarg","called after endarg");
	}
	
	if ((FLAGS & INIT) == 0)
	{
		return ga_getarg_err("countarg","not initialized with setarg");
	}
	

	if (NLIST == 0 || ARGLIST == NULL) 
	{
		return (0);
	}
	

	found=0;

	h = ga_compute_hash(name);

	/*  
	 *  if list is NULL then return NOW; the following "for" loop
	 *  the pointer ARGLIST + (NLIST-1) gives ARGLIST -1 which
	 *  does not test to < ARGHEAD for some reason
         */
	
	if (NLIST <= 0)
	{ 
		return (0);
	}
	

	/* search list backwards, stopping at first find */
	for (alptr = ARGLIST +(NLIST-1); alptr >= ARGLIST; alptr--)
	{
		if (alptr->hash != h) 
			continue;
		if (strcmp(alptr->argname, name) != 0) 
			continue;
		str = alptr->argval;
		ptr = str;
		
		switch (*type)
		{
	       	case 'd':
		case 'f':
		case 'F':
			/*
			 * Count the number of commas to detemine list
			 * size.  If the str isn't NULL than there
			 * is at least one element and at least one
			 * more than there are commas.
			 */
			while ((ptr = strchr(ptr, ',')) != NULL)
			{
					found++;
					ptr++;
			}
			if (str != NULL)
				found++;
			break;
	       	case 's':
			while ((ptr = strchr(ptr, ',')) != NULL)
			{
				if (ptr[-1] != '\\')
					found++;
				ptr++;
			}
			if (str != NULL)
				found++;
			break;
	       	default:
		       	return ga_getarg_err("countarg",
		       		      "unknown conversion type %s",type);
		       	break;
		}
		break;
	}
	return (found);
}	

int lenarg(char *name)
{
	int 	h, len, new_len;
	char 	*str, *ptr1, *ptr2;
	register struct arglist *alptr;
	
	if (FLAGS & END_PAR)
	{
		return ga_getarg_err("lenarg","called after endarg");
	}
	
	if ((FLAGS & INIT) == 0)
	{
		return ga_getarg_err("lenarg","not initialized with setarg");
	}

	if (NLIST == 0 || ARGLIST == NULL) 
	{
		return (0);
	}
	

	h = ga_compute_hash(name);

	/*  
	 *  if list is NULL then return NOW; in the following "for"
	 *  loop the pointer ARGLIST + (NLIST-1) gives ARGLIST -1
	 *  which does not test to < ARGHEAD for some reason
         */
	
	if (NLIST <= 0)
	{ 
		return (0);
	}
	

	/* search list backwards, stopping at first find */
	for (alptr = ARGLIST +(NLIST-1); alptr >= ARGLIST; alptr--)
	{
		if (alptr->hash != h) 
			continue;
		if (strcmp(alptr->argname, name) != 0) 
			continue;
		str = alptr->argval;
		ptr1 = str;
		len = new_len = 0;
	       	while ((ptr2 = strchr(ptr1, ',')) != NULL)
	       	{
	       		new_len += ptr2 - ptr1;
			if (ptr2[-1] != '\\')
			{
				if (new_len > len)
					len = new_len;
				new_len = 0;
			}
			ptr2++;
			ptr1 = ptr2;
		}
			
		if (ptr1)
			new_len = strlen(ptr1);
		if (new_len > len)
			len = new_len;
		
		break;
	}
	return (len);
}	



int
#ifdef FORTRAN
fgtarg_(char *name, char *type, void *val, int dum1, int dum2, int lens)
/* dum1 & dum2 are extra args that fortran puts in */
#else
getarg(char *name, char *type, void *val)
#endif
   {
	register char *sptr;
	register struct arglist *alptr;
	double *dbl;
	float *flt;
	int *intr;
	int h, hno, hyes, found;
	char noname[MAXNAME+2], *str, *ptr;

	/*fprintf(stderr,"looking for %s, type=%s\n",name,type);*/
	if(FLAGS & END_PAR)
		return ga_getarg_err("getarg","called after endarg");
	if( (FLAGS & INIT) == 0)
		return ga_getarg_err("getarg","not initialized with setarg");
	if (val == NULL)
		return ga_getarg_err("getarg", "NULL pointer value");
	
	if(NLIST == 0 || ARGLIST == NULL) return(0);

	/* The following line corrects a common input error */
	if(type[1]=='v') { type[1]= type[0]; type[0]='v'; }

	found=0;

	if(*type == 'b') goto boolean;

	h= ga_compute_hash(name);

	/*  
	 *  if list is NULL then return NOW; the following "for" loop
	 *  the pointer ARGLIST + (NLIST-1) gives ARGLIST -1 which
	 *  does not test to < ARGHEAD for some reason
         */
	
	if(NLIST <= 0) return(found);


	/* search list backwards, stopping at first find */
	for(alptr= ARGLIST +(NLIST-1); alptr >= ARGLIST; alptr--)
	   {
		/*fprintf(stderr,"getarg: checking %s\n",alptr->argname);*/
		if(alptr->hash != h) continue;
		if(strcmp(alptr->argname,name) != 0) continue;
		str= alptr->argval;
		switch(*type)
		   {
			case 'd':
			        intr= (int *) val;
				*intr= (int) strtol(str, &ptr, 0);
				if (ptr == str)
					found = 0;
				else
					found = 1;
				break;
			case 'f':
				flt= (float *) val;
				*flt= (float) strtod(str, &ptr);
				if (ptr == str)
					found = 0;
				else
					found = 1;
				break;
			case 'F':
				dbl= (double *) val;
				*dbl= strtod(str, &ptr);
				if (ptr == str)
					found = 0;
				else
					found = 1;
				break;
			case 's':
                                sptr= (char *) val;
                                while(*str) *sptr++ = *str++;
#ifdef FORTRAN
				while(sptr < (((char *)val)+lens)) 
					*sptr++ = ' ';
#else
                                *sptr= '\0';
#endif
                                found=1;
                                break;
			case 'v':
/* 
 *  For now, it's an error to get a Fortran vector of strings,
 *  If ga_getvector returns an error, stop processing and
 *  return GETARG_ERROR (-1)
 */

#ifdef FORTRAN
				if (type[1] == 's')
					return ga_getarg_err("getarg",
							"fortran string vector");
				else
					if ((found = ga_getvector(str, type, val))
					    == GETARG_ERROR)
						return found;
#else
				if ((found = ga_getvector(str,type,val)) ==
				    GETARG_ERROR)
					return found;
#endif
				break;
			default:
				return ga_getarg_err("getarg",
					"unknown conversion type %s",type);
				break;
		   }
		break;
	   }
	return(found);
boolean:

	/*  
	 *  if list is NULL then return NOW; the following "for" loop
	 *  the pointer ARGLIST + (NLIST-1) gives ARGLIST -1 which
	 *  does not test to < ARGHEAD for some reason
         */
	
	if(NLIST <= 0) return(found);

	sprintf(noname,"no%s",name);
	hno = ga_compute_hash(noname);
	hyes= ga_compute_hash(  name);
	found=0;
	/* search list backwards, stopping at first find */
	for(alptr= ARGLIST +(NLIST-1); alptr >= ARGLIST; alptr--)
	   {
		if(alptr->hash != hno && alptr->hash != hyes) continue;
		if(strcmp(alptr->argname,  name)== 0)
		   {
			if(alptr->argval[0] == '\0')
			  *( (int *) val)= 1;
			else
			  *( (int *) val)= atol(alptr->argval);
			found++;
			break;
		   }
		if(strcmp(alptr->argname,noname)== 0)
		   {	*( (int *) val)= 0; found++; break; }
	   }
	return(found);
   }

int ga_compute_hash(char *s)
   {
	register int h;
	h= s[0];
	if(s[1]) h |= (s[1])<<8;	else return(h);
	if(s[2]) h |= (s[2])<<16;	else return(h);
	if(s[3]) h |= (s[3])<<24;
	return(h);
   }

int ga_getvector(char *list, char *type, void *val)
   {
	register char *p;
	register int index, cnt;
	char *valptr, **strptr, *svalptr, sval[MAXVALUE], *ptr;
	int limit;
	int ival, *iptr;
	float fval, *fptr;
	double dval, *dptr;

	limit= MAXVECTOR;
	if(type[2] == '(' || type[2] == '[') limit= atol(&type[3]);
	if(limit <= 0)
		return ga_getarg_err("getarg","bad limit=%d specified",limit);
	/*fprintf(stderr,"limit=%d\n",limit);*/
	index= 0;
	p= list;
	while(*p != '\0'  && index < limit)
	   {
		cnt=1;
	 backup: /* return to here if we find a repetition factor */
		while(*p == ' ' || *p == '\t') p++;
		if(*p == '\0') return(index);
		valptr= p;
	getvalue: /* return here if valid value in char*[] arg */
		while( *p != ',' && *p != '*' && *p != 'x' && *p != 'X' &&
			*p != '\0') p++;
		if (type[1] == 's' && ((*p == ',' && p[-1] == '\\')
				       || *p == '*' || *p == 'x' || *p == 'X'))
		{	p++; goto getvalue; }
		if((*p == '*' || *p == 'x' || *p == 'X') && 
		   type[1] != 's')
		   {
			cnt= atol(valptr);
			if(cnt <= 0)
				return ga_getarg_err("getarg",
					"bad repetition factor=%d specified",
					 cnt);
			if(index+cnt > limit) cnt= limit - index;
			p++;
			goto backup;
		   }
		/*fprintf(stderr,"index=%d cnt=%d p=%s$\n",index,cnt,p);*/
		switch(type[1])
		   {
			case 'd':
				iptr= (int *) val;
				ival= (int) strtol(valptr, &ptr, 0);
				if (ptr == valptr)
					index = 0;
				else
				{
					if (iptr+index == NULL)
						return ga_getarg_err("getarg",
			       		          "NULL vector ptr at index %d.", index);
					while(cnt--)
					       	iptr[index++] = ival;
				}
				break;
			case 'f':
				fptr= (float *) val;
				fval= (float) strtod(valptr, &ptr);
				if (ptr == valptr)
					index = 0;
				else
				{
					if (fptr+index == NULL)
						return ga_getarg_err("getarg",
				       	          "NULL vector ptr at index %d", index);
					while(cnt--) 
						fptr[index++] = fval;
				}
				break;
			case 'F':
				dptr= (double *) val;
				dval= strtod(valptr, &ptr);
				if (ptr == valptr)
					index = 0;
				else
				{
					if (dptr+index == NULL)
						return ga_getarg_err("getarg",
				       	          "NULL vector ptr at index %d", index);
					while(cnt--) 
						dptr[index++] = dval;
				}
				break;
			case 's':
				svalptr = sval;
				strptr = (char **) val;
				*svalptr = '\0';
				while (*valptr != '\'' && *valptr != '\0')
				{
					if (*valptr == '\\')
						valptr++;
				       	*svalptr++ = *valptr++;
				}
                                *svalptr= '\0';
				if (strptr+index == NULL)
					return ga_getarg_err("getarg",
		       	       		  "NULL vector ptr at index %d", index);
				strcpy(strptr[index++], sval);
				if (*p != '\0') p++;
				break;
			default:
				return ga_getarg_err("getarg",
					"bad vector type=%c specified",type[1]);
				break;
		   }
		/* if conversion couldn't be made return 0 */
		if (index == 0) 
			break;
		if (*p != '\0') 
			p++;
	   }
	return(index);
   }

int
ga_getarg_err(char *subname, char *format, ...)
   {
        va_list ap;
	va_start(ap, format);
	(void) fprintf(stderr,"\n***** ERROR in %s[%s] *****\n\t",
		(PROGNAME == NULL ? "(unknown)" : PROGNAME),subname);
	(void) vfprintf(stderr, format, ap);
	va_end(ap);
	(void) fprintf(stderr,"\n");
	return GETARG_ERROR;
   }
#ifndef FORTRAN
char *getsarg(char *parname, char *defvalue)
{
	char parvalue[1024];

	parvalue[0] = '\0';

	if(defvalue)
	{
		strcpy(parvalue, defvalue);
	}
	
	getarg(parname, "s", parvalue);

	if(! defvalue && parvalue[0] == '\0') return( (char *) 0 );

	else return(strdup(parvalue));
}
#endif
